### DISCUSSION TITLE
Iterate over trial dictionaries stored in an array

### DISCUSSION THREAD

**Discussion by gaiamolinaro** [jspsych/jsPsych#970] [Q&A] [answered]
Hi all! I am developing a task where I need to iterate over trials stored in an array called `trials` 
Each `trial` in `trials` is a dictionary that contains a number of key-values pairs, for instance:
```
var trial_1 = {
  stimulus_left: [image_left_1, image_left_2], 
  stimulus_right: [image_right_1, image_right_2]
};
```
I pushed each of the trials I coded in this way in the `trials` array. 
My timeline is written as follows, where `num_games` is the number of trials (let's say I have defined `var num_games = 10`. At the end, I use `timeline: timeline`
```
for (var i = 0; i < num_games; i++) { 
    // experiment timeline repeated for num_games
     var node = {
         timeline: [
 {//here goes my code for what to show in the first screen},
{//here is what to show next, etc. }
    ]
         } 
     timeline.push(node) 
         }
```
Within the experiment timeline, I use `trial.stimulus_left` to refer, for instance, to the trial that I want to show on the left. I need `trial` to be different every time. Ideally, I need to put a variable `trial = trials[i]` somewhere in my code, but every attempt to do so fails for one reason or another (I tried putting it right after the for loop, inside my timeline as a dictionary key and variable, and inside the "stimulus" parameter (as a function) for my first screen). I could also directly use something like `for trial of trials` (if that's the equivalent of Python's `for x in my_array`) as my for loop iterator, but I am quite new to Javascript and I have failed to do so at the moment. 
Thank you so much for your time and help!

**Comment by becky-gilbert** [comment, marked as answer]
Hi,

Just to check - your `trials` array is an array of objects, where each object contains the information for a single trial, like this?:
```javascript
var trials = [
 { // trial 1
  stimulus_left: [image_left_1, image_left_2], 
  stimulus_right: [image_right_1, image_right_2]
 },
 { // trial 2
  stimulus_left: [image_left_1, image_left_2], 
  stimulus_right: [image_right_1, image_right_2]
 }
 ...
];
```
Is that right?
And if `num_games` is equal to the number of trials, then it should also be equal to `trials.length`, right?

For your loop, it would help to know a little bit more about what plugin(s) you're using to set up the individual trials and how you're using the information stored in your `trials` array. It doesn't look like your objects inside of the `trials` array have key-value pairs that directly map onto jsPsych trial parameters, so I'm not clear on how the values will be used in a jsPsych trial object. So the easiest thing might be to change your `trials` array (if necessary) so that each object maps neatly onto jsPsych trial parameters.

Nonetheless, you should be able to dynamically create trial objects using the information in each object in your `trials` array. Here's an example:
```javascript
for (var i = 0; i < num_games; i++) { 
    // this assumes that num_games = trials.length
    timeline.push(
        {
            type: 'image-button-response',
            stimulus: trials[i].stimulus_left[0]
        },
        {
            type: 'image-button-response',
            stimulus: trials[i].stimulus_left[1]
        },
        {
            type: 'image-button-response',
            stimulus: trials[i].stimulus_right[0]
        },

        {
            type: 'image-button-response',
            stimulus: trials[i].stimulus_right[1]
        },
    );
}
```
Or if you want to use timeline variables, then you could do something like this:
```javascript
for (var i = 0; i < num_games; i++) { 
    var node = {
        timeline: [
            {
                type: 'image-button-response',
                stimulus: jsPsych.timelineVariable('stim_left_1')
            },
            {
                type: 'image-button-response',
                stimulus: jsPsych.timelineVariable('stim_left_2')
            },
            {
                type: 'image-button-response',
                stimulus: jsPsych.timelineVariable('stim_right_1')
            },

            {
                type: 'image-button-response',
                stimulus: jsPsych.timelineVariable('stim_right_2')
            }
        ],
        timeline_variables: [
            {stim_left_1: trials[i].stimulus_left[0], stim_left_2: trials[i].stimulus_left[1], stim_right_1: trials[i].stimulus_right[0], stim_right_2: trials[i].stimulus_right[1]}
        ]
    }
    timeline.push(node);
}
```

**Reply by gaiamolinaro** [reply]
Hey Becky, thank you so much for your reply! As for your questions, my `trials` array is an array of objects, similar to what you have described, except I am defining each as a variable (var) and then push each into `trials` - would that be a problem? 

`num_games` may or may not be equal to the number of trials; I may want not to go over every single trial but just sample randomly, or iterate over one more than once. But let's keep things simple for the moment and assume that, yes, they are the same length. 

I am using mostly` html-button-response `and `hmtl-keyboard-response` plugins. My stimuli are typically a background with some items appearing on it; then, I want to make some items stay or disappear depending on the participant's responses; I also need a random component into which specific items stay and which one disappear, so I need some dynamicity. What I need is to refer to values stored as keys in each trial so that each trial is consistent with other values in it, but also general enough to deal with it. So with the example above, I would have:
```
var trial_1 = {
  stimulus_left: [image_left_1a, image_left_2b], 
  stimulus_right: [image_right_1a, image_right_2b]
};
trials.push(trial_1);
var trial_2 = {
  stimulus_left: [image_left_1c, image_left_2d], 
  stimulus_right: [image_right_1c, image_right_2d]
};
trials.push(trial_2);
```
And then, in the timeline, perhaps something like:
```
 {
        type: 'html-keyboard-response',
		stimulus: "<div class='container'> <img class='background' src='" + Background + "'>" + trial.stimulus_left + trial.stimulus_right + "</div>",
        choices: jsPsych.NO_KEYS,
        trial_duration: 4000
        },
```
And each time, I would have `trial = trials[0]`, then `trials = trials[1]`, and so on. 
Did this clarify my question?

**Reply by jodeleeuw** [reply]
Hi @gaiamolinaro ,

This sounds like the typical use-case for [timeline variables](https://www.jspsych.org/overview/timeline/#timeline-variables). Have you taken a look at the [tutorial content that covers how to use them](https://www.jspsych.org/tutorials/rt-task/#part-5-timeline-variables)?

**Reply by becky-gilbert** [reply]
Hi @gaiamolinaro, thanks for clarifying. 

There are a few issues in your trial object code above. One is that the image source is set to `Background`, which isn't defined in the code snippet. Is `Background` a variable that contains a path to an image file? E.g. 
`var Background = "img/background.png"`

Another issue is that, after your image element, you are putting in the value of `trial.stimulus_left` and `trial.stimulus_left`. The value of those two keys are arrays with 2 elements, so you'll need to index those arrays with either [0] or [1] to get the specific values that you're after. And I'm assuming that the elements inside these arrays are actually paths to image files, e.g. `image_left_1a` is something like `"img/left_1a.png"`. If so, then you'll want to add corresponding `<img>` tags for each of those images, and use the value of the array element as the image `src`. In the example below, I added a left and right image as you did above, but just selected the first image from each array.
```javascript
for (var i = 0; i < num_games; i++) { 
    var trial = trials[i];
    timeline.push(
        {
            type: 'html-keyboard-response',
	    stimulus: "<div class='container'> <img class='background' src='" + trial.stimulus_left[0] + "'><img class='background' src='" + trial.stimulus_right[0] + "'></div>",
            choices: jsPsych.NO_KEYS,
            trial_duration: 4000
        }
    );
}
```
Does this help at all? Sorry if I've again misunderstood what you're trying to do!

**Reply by becky-gilbert** [reply]
And just to elaborate on Josh's comment about using timeline variables instead of a loop - you could construct your `trials` array so that the keys in each object correspond to one parameter value in your trial objects. For example, you could have separate keys corresponding to each image file (rather than an array with multiple images). Like this:
```javascript
var trials = [
 { // trial 1
  stimulus_left_1: image_left_1a, 
  stimulus_left_2: image_left_2b, 
  stimulus_right_1: image_right_1a, 
  stimulus_right_2: image_right_2b
 },
 { // trial 2
  stimulus_left_1: image_left_1c, 
  stimulus_left_2: image_left_2d, 
  stimulus_right_1: image_right_1c, 
  stimulus_right_2: image_right_2d
 }
 ...
];
```
Then, instead of constructing your trials in a loop, you'd be able to use timeline variables like this:
```javascript
var trial_procedure = {
  timeline: [
    {
      type: 'html-keyboard-response',
      stimulus: function() {
        return "<div class='container'><img class='background' src='" + jsPsych.timelineVariable('stimulus_left_1', true) + "'><img class='background' src='" +  jsPsych.timelineVariable('stimulus_right_1', true) + "'></div>"
      },
      choices: jsPsych.NO_KEYS,
      trial_duration: 4000
    },
    {
      type: 'html-keyboard-response',
      stimulus: function() {
        return "<div class='container'><img class='background' src='" + jsPsych.timelineVariable('stimulus_left_2', true) + "'><img class='background' src='" +  jsPsych.timelineVariable('stimulus_right_2', true) + "'></div>"
      },
      choices: jsPsych.NO_KEYS,
      trial_duration: 4000
    }
  ],
  timeline_variables: trials
};
```
Note that now your stimulus parameter has to be a function because you're combining timeline_variables with HTML to create your stimulus value, like in the second example from the documentation: https://www.jspsych.org/overview/timeline/#timeline-variables

**Reply by gaiamolinaro** [reply]
Thanks for directing me to that page, @jodeleeuw!
And thank you again, @becky-gilbert , your latest comment actually seems like it could help a lot. My only problem with that is that some of my keys actually correspond to many image variables (yes, linking to image paths as you said above for background), which I often want to join in a single HTML line. And sometimes I want to only use three out of the six images I have for one parameter. So if I had:
var trials = [
```
 { // trial 1
  stimulus_left_1: [image_left_1a, image_left_1b,  image_left_1c],
  stimulus_left_2: image_left_2a
  stimulus_right_1: [image_right_1a, image_right_1b, image_right_1c],
  stimulus_right_2: image_right_2a
 },
 { // trial 2
  stimulus_left_1: [image_left_1a, image_left_1b,  image_left_1c],
  stimulus_left_2: image_left_2a
  stimulus_right_1: [image_right_1a, image_right_1b, image_right_1c],
  stimulus_right_2: image_right_2
 }
 ...
];
```
I would like to play around with it and sometimes display e.g all of the images for stimulus_left_1 (using the join.(" ") method), and other times only show the first two. 
I tried storing
  `var simulus_for_example = "<img class='example_class' src='" +  jsPsych.timelineVariable('stimulus_eft_1', true) + "'>";`
as a variable, but it didn't work. 
Do you have any clue on how to solve this? Sorry if my explanation isn't clear again. I'd be happy to clarify wherever needed!

**Reply by becky-gilbert** [reply]
In that case, one thing you could do is have a variable in your trial data that tells you how many images to display, and use that value in your stimulus function to change the way that the stimulus string is constructed. 
For example, your trial data might look like this:
```javascript
var trials = [
{ // trial 1
  n_images_left: 3,
  stimulus_left_1: [image_left_1a, image_left_1b,  image_left_1c],
},
{ // trial 2
  n_images_left: 2,
  stimulus_left_1: [image_left_1a, image_left_1b,  image_left_1c],
 }
 ...
];
```
And your trial object might look like this:
```javascript
{
  type: 'html-keyboard-response',
  stimulus: function() {
    var images = jsPsych.timelineVariable('stimulus_left_1', true);
    var n_images_left = jsPsych.timelineVariable('n_images_left', true);
    var stim_string = "<div class='container'>";
    for (var i=0; i<n_images_left; i++) {
      stim_string += "<img id='left" + i.toString() + "' src='" + images[0] + "'>";
    }
    stim_string += "</div>";
    return stim_string;
  },
  choices: jsPsych.NO_KEYS,
  trial_duration: 4000
}
```
Or, if you don't know in advance how many images to show for each trial, then you could determine that information some other way inside of your stimulus function (instead of storing it in the trial data). For instance, you could create a global `n_images_left` variable and change its value during the experiment. Or you could access the jsPsych data stored so far and use that to determine the number of images to show.

Does that help with what you're trying to do?

**Reply by gaiamolinaro** [reply]
Yes, thank you so much once again!
